### PR TITLE
Convert backend to TypeScript (50% -> 64%)

### PR DESCRIPTION
**Pull request by eliandoran**
Language|files|blank|comment|code
:-------|-------:|-------:|-------:|-------:
TypeScript|138|3828|1117|15036
JavaScript|106|2333|869|8469
--------|--------|--------|--------|--------
SUM:|244|6161|1986|23505

### CODE REVIEW

File: src/services/meta/note_meta.ts [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
-    notePath: string;
-    isClone: boolean;
-    title: string;
-    notePosition: number;
-    prefix: string;
-    isExpanded: boolean;
-    type: string;
-    mime: string;
+    noteId?: string;
+    notePath?: string[];
```
Comment (alexpietsch) [inline src/services/meta/note_meta.ts:6]: Why was this changed to an array?

File: src/services/export/zip.ts [hunk 2]
```diff
@@ -136,24 +112,26 @@ async function exportToZip(taskContext, branch, format, res, setHeaders = true)
...
-            const meta = new NoteMeta();
-            meta.isClone = true;
-            meta.noteId = note.noteId;
-            meta.notePath = notePath;
-            meta.title = note.getTitleOrProtected();
-            meta.prefix = branch.prefix;
-            meta.dataFileName = fileName;
-            meta.type = 'text'; // export will have text description
-            meta.format = format;
+            const meta: NoteMeta = {
```
[the remark this answers is not in the document]
Comment (eliandoran) [reply src/services/export/zip.ts:121]: @alexpietsch , yes, let's keep it as the original for now.

File: src/services/meta/note_meta.ts (hunk 1 above)
Comment (eliandoran) [reply src/services/meta/note_meta.ts:6]: @alexpietsch , I believe the original data type (from my previous PR) was incorrect. Note paths are expressed as an array of strings.

**Comment by eliandoran** [comment]
@alexpietsch, please re-review.

File: src/services/meta/note_meta.ts (hunk 1 above)
Comment (alexpietsch) [reply src/services/meta/note_meta.ts:6]: I tested it, it is actually an array.

**Review by alexpietsch — APPROVED** [review]

File: src/services/import/zip.ts [hunk 3]
```diff
@@ -1,43 +1,45 @@
...
-    /** @type {BNote} */
-    let firstNote = null;
-    /** @type {Set.<string>} */
-    const createdNoteIds = new Set();
-
-    function getNewNoteId(origNoteId) {
+    /** path => noteId | attachmentId */
+    const createdPaths: Record<string, string> = { '/': importRootNote.noteId, '\\': importRootNote.noteId };
+    let metaFile!: MetaFile;
+    let firstNote!: BNote;
```
Comment (j-tirelli) [inline src/services/import/zip.ts:39]: Do we gain any value using the definite assignment operator here, vs the type being null or MetaFile?

IE:
```suggestion
    let metaFile!: MetaFile | null = null;
    let firstNote!: BNote | null = null;
```

File: src/services/import/zip.ts [hunk 4]
```diff
@@ -88,30 +86,30 @@ async function importZip(taskContext, fileBuffer, importRootNote) {
...
+        let cursor: NoteMeta | undefined = {
             isImportRoot: true,
-            children: metaFile.files
+            children: metaFile.files,
+            dataFileName: ""
         };
 
-        /** @type {NoteMeta} */
-        let parent;
+        let parent!: NoteMeta;
```
Comment (j-tirelli) [inline src/services/import/zip.ts:102]: Same here, I'm curious what the benefit is over:
```suggestion
        let parent: NoteMeta | undefined;
```

File: src/services/import/markdown.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+"use strict";
+
+import marked = require("marked");
+import htmlSanitizer = require('../html_sanitizer');
+import importUtils = require('./utils');
+
+function renderToHtml(content: string, title: string) {
+    const html = marked.parse(content, {
+        async: false
+    }) as string;  // FIXME: mangle and headerIds does not seem to exist in marked
```
[the remark this answers is not in the document]
Comment (eliandoran) [reply src/services/import/markdown.ts:10]: @j-tirelli , after some digging I was able to identify them in the documentation of `marked` for [options](https://marked.js.org/using_advanced#options).

`mangle` was set by default to `true` and now says:

> Removed in v8.0.0 use [marked-mangle](https://www.npmjs.com/package/marked-mangle) to mangle email addresses.

`headerIds` was also set by default to `true`:

> Removed in v8.0.0 use [marked-gfm-heading-id](https://www.npmjs.com/package/marked-gfm-heading-id) to include an id attribute when emitting headings (h1, h2, h3, etc).

zadam added `marked` back in [July 2023](https://github.com/zadam/trilium/commit/64b86b2), but with a pretty old version for some reason (5.1.1). Since then it was updated multiple times, so probably it just ended up as technical debt.

Since we now know what's the reason behind them, we can safely remove the TODO.

File: src/services/import/zip.ts (hunk 3 above)
Comment (eliandoran) [reply src/services/import/zip.ts:39]: Seems to work with an extra null check at the end of the `importZip` method.

File: src/services/import/zip.ts (hunk 4 above)
Comment (eliandoran) [reply src/services/import/zip.ts:102]: @j-tirelli , I don't think there's any benefit to using the `!` operator. I used it to avoid the errors about using a value before it's declared. I have to admit the code in `zip.ts` was a bit hard to follow for me, so I just did what was necessary to avoid any build errors. Nevertheless, the `| null = null` approach is much cleaner since it provides better type checking, even if it requires adding some new null checks here and there.

**Comment by eliandoran** [comment]
@j-tirelli , thank you for the review. Please re-review when you have the time.

**Review by j-tirelli — APPROVED** [review]
Changes look good to me.
